File: ripperdoc/dock/idea/toolbox/open_url.py
#!/usr/bin/env python3
"""Open an http/https URL in the running JetBrains IDE's embedded JCEF preview."""

import argparse
import sys
from urllib.parse import urlsplit

from core import IdeaError, in_idea, preview_url, resolve_exec_path


def open_url(url: str, title: str | None = None) -> None:
    """Open `url` in the running IDE's embedded JCEF web-preview tab.

    FIRE-AND-FORGET: there is no editor to block on, so unlike open_file there is
    no `--wait`. The tab is titled "Preview of <title>"; `title` defaults to the
    URL's host[:port] (see preview_url).

    Clean-fail, NOT a browser fallback: resolve_exec_path() is the single guard
    for a live IDE — it raises IdeaError (no IDE in the ancestry) or
    NotImplementedError (non-macOS stub), which the CLI turns into a non-zero
    exit. Because the IDE preview is the only thing this command can do, there is
    deliberately no shell-out to `open`/a browser when it's unavailable. With a
    live IDE confirmed, preview_url() fires the ideScript (the in-IDE registry
    gate handles a JCEF-off WebStorm as a clean no-op).
    """
    resolve_exec_path()
    preview_url(url, title)


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        prog="open_url.py",
        description="Open an http/https URL in the running JetBrains IDE's embedded JCEF preview.",
        epilog=(
            "Examples:\n"
            "  open_url.py http://localhost:3000              # preview a local dev server\n"
            "  open_url.py https://example.com --title docs   # custom tab label"
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("url", help="http/https URL to open in the embedded preview")
    parser.add_argument("--title", default=None, help="tab label (default the URL's host[:port])")
    ns = parser.parse_args(argv)
    # Light validation: a non-empty http/https URL. The IDE's JCEF preview only
    # speaks http(s), so reject anything else up front with a clear note.
    if urlsplit(ns.url).scheme not in {"http", "https"}:
        print("open_url: url must be a non-empty http/https URL", file=sys.stderr)
        return 1
    try:
        # Cheap CLI gate: fail fast/clean outside a JetBrains terminal, before
        # resolve_exec_path()'s deeper ancestry walk. Reuse the IdeaError
        # path so the message is identical to the resolver-triggered failure.
        if not in_idea():
            raise IdeaError("no JetBrains IDE in the process ancestry")
        open_url(ns.url, ns.title)
    except (IdeaError, NotImplementedError) as exc:
        print(f"open_url: {exc}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: ripperdoc/dock/idea/toolbox/open-url.ts
#!/usr/bin/env bun
/**
 * open-url.ts — open an http/https URL in the running JetBrains IDE's embedded
 * JCEF preview.
 *
 * FIRE-AND-FORGET: there is no editor to block on, so unlike open-file there is
 * no --wait. Clean-fail, NOT a browser fallback: resolveExecPath() is the single
 * guard for a live IDE — it throws IdeaError / NotImplementedError; the CLI turns
 * either into a non-zero exit. With a live IDE confirmed, previewUrl() fires the
 * ideScript.
 *
 * The CLI is a cmdore commandless command: cmdore owns parsing, help, the global
 * flags (--quiet/--verbose/--json/--dry-run/--help/--version), and exit codes.
 * main() wraps execute() with onError:"throw" so it can keep the repo CLI shape
 * (return a number; process.exit only under the import.meta.main guard) and map
 * IdeaError / CmdoreError to the open-url: prefixed stderr line.
 */

import { CmdoreError, defineCommand, effect, execute } from "cmdore"
import { parseUrl } from "../../../../lib/text.ts"
import { IdeaError, NotImplementedError } from "./core/errors.ts"
import { inIdea, previewUrl as rawPreviewUrl, resolveExecPath } from "./core/index.ts"

const PROG = "open-url"

/** cmdore metadata for the commandless CLI; version mirrors the idea plugin manifest. */
const METADATA = {
    name: PROG,
    version: "0.1.0",
    description: "Open an http/https URL in the running JetBrains IDE's web preview."
} as const

/**
 * The write side-effect, wrapped as cmdore `effect.fn` so it is skipped on
 * `--dry-run` (when cmdore flips `effect.enabled` off) and mockable in tests by
 * the WRAPPER REFERENCE (`effect.mock(previewUrl, …)`) — no per-file mutable
 * seam. `resolveExecPath` is the live-IDE READ guard and stays unwrapped.
 */
export const previewUrl = effect.fn(rawPreviewUrl, "ide.previewUrl")

/**
 * Open `url` in the running IDE's embedded JCEF web-preview tab. resolveExecPath()
 * is the single guard for a live IDE; then previewUrl() fires the ideScript.
 */
export const openUrl = async (url: string, title?: string): Promise<void> => {
    await resolveExecPath()
    await previewUrl(url, title)
}

/**
 * The cmdore command behind the CLI. Validates the http(s) URL, gates on a live
 * IDE (cheap fail-fast before resolveExecPath()'s deeper ancestry walk), then
 * runs openUrl.
 */
const openUrlCommand = defineCommand({
    name: PROG,
    description: METADATA.description,
    arguments: [{ name: "url", description: "http/https URL to open", required: true }],
    options: [{ name: "title", arity: 1, hint: "title", description: "title for the preview tab" }],
    run: async ({ url, title }) => {
        // Light validation: a non-empty http/https URL. The IDE's JCEF preview only
        // speaks http(s), so reject anything else up front with a clear note.
        if (!["http", "https"].includes(parseUrl(url).scheme)) {
            throw new IdeaError("url must be a non-empty http/https URL")
        }
        if (!inIdea()) {
            throw new IdeaError("no JetBrains IDE in the process ancestry")
        }
        await openUrl(url, title)
    }
})

/**
 * CLI entrypoint. Hands argv to cmdore (parsing, help, global flags), then maps
 * the domain failures to the open-url: stderr line and their exit codes:
 * IdeaError / NotImplementedError -> 1, CmdoreError (missing/unknown arg) -> its
 * own exitCode. Anything else is a bug and rethrown.
 */
export const main = async (argv: string[] = Bun.argv.slice(2)): Promise<number> => {
    try {
        await execute(openUrlCommand, { argv, metadata: METADATA, onError: "throw" })
    } catch (error) {
        if (error instanceof IdeaError || error instanceof NotImplementedError) {
            process.stderr.write(`${PROG}: ${error.message}\n`)
            return 1
        }
        if (error instanceof CmdoreError) {
            process.stderr.write(`${PROG}: ${error.message}\n`)
            return error.exitCode
        }
        throw error
    }
    return 0
}

if (import.meta.main) {
    process.exit(await main())
}
